---
title: "InVitro_Analysis_090625"
author: "Winnie Lei"
date: "2025-06-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load library}
setwd(".")
library(edgeR)
library(knitr)
library(ggplot2)
library(colorspace)
library(ggh4x)
library(ggpubr)
library(rstatix)
library(limma)
library(ExperimentHub)
library(ggrepel)
library(clusterProfiler)
library(org.Hs.eg.db)
```

# Analysis of In Vitro RNAseq data

## 1. Preparation of data and environment

```{r Read in RNAseq data}
cell_count = read.csv("./Invitro_cellcount.csv", row.names = 1)
```

Read in sample data
```{r Read in sample data}
sample_data = read.csv("InVitro_Sampledata.csv")
sample_data
```

Calculate normalized data

```{r Normalize RNAseq data}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = cell_count,
                              colData = sample_data,
                              design = ~ Condition)
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
logTMM = log10(normalized_counts+1)
```

## 2. Perform differential expression analysis

```{r data preprocessing}
d0 <- DGEList(cell_count)
d0 <- calcNormFactors(d0)
cutoff <- 1
drop <- which(apply(edgeR::cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 
```

```{r Define group}
condition_dict = setNames(sample_data$Condition,sample_data$File_name)
group = unname(condition_dict[colnames(d0)])
group
```

```{r Voom transformation}
mm <- model.matrix(~0 + group)
y <- voom(d, mm, plot = F)
```

```{r Fit linear models in limma}
fit <- lmFit(y, mm)
contr <- makeContrasts(groupDKO - groupWT, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
top.table <- topTable(tmp, sort.by = "P", n = Inf)
head(top.table, 20)
```

Number of DE genes
```{r Number of DE genes}
length(which(top.table$adj.P.Val < 0.05))
```

```{r Write top table}
top.table$Gene <- rownames(top.table)
top.table <- top.table[,c("Gene", names(top.table)[1:6])]
write.csv(top.table, 
            file = "../res/DKO_WT_DEGs.csv", 
            row.names = F,
            quote = F)
```

```{r Add labels for differential expression}
# add a column of NAs
top.table$diffexpressed <- "NO"
# if log2Foldchange > 1 and pvalue < 0.05, set as "UP" 
top.table$diffexpressed[top.table$logFC >= 1.5 & top.table$P.Value < 0.01] <- "UP"
# if log2Foldchange < -1 and pvalue < 0.05, set as "DOWN"
top.table$diffexpressed[top.table$logFC <= -1.5 & top.table$P.Value  < 0.01] <- "DOWN"
top.table$delabel <- NA
top.table$delabel[top.table$diffexpressed != "NO"] <- rownames(top.table[top.table$diffexpressed != "NO",])
top.table$delabel = ifelse(top.table$delabel=="B2M","B2M",ifelse(top.table$delabel=="RFX5","RFX5",NA))
```

```{r Visualize differenial expression results}
ggplot(data=top.table, 
       aes(x=logFC, y=-log10(P.Value), col=diffexpressed, label=delabel)) +
   geom_point() + 
  theme_minimal() +
  geom_text_repel() +
  scale_color_manual(values=c("blue", "grey", "red")) +
  geom_vline(xintercept=c(-1.5, 1.5), col="red") +
  geom_hline(yintercept=-log10(0.01), col="red")
```

Perform pathway enrichment analysis
```{r Pathway enrichment analysis }
DEG_Input <- top.table$logFC
names(DEG_Input) <- rownames(top.table)
DEG_Input <-na.omit(DEG_Input )
DEG_Input  = sort(DEG_Input , decreasing = TRUE)

GSEA_result <- gseGO(geneList=DEG_Input,
                 ont ="ALL",
                 keyType = "ALIAS",
                 minGSSize = 3,
                 nPermSimple = 10000,
                 maxGSSize = 800,
                 pvalueCutoff = 0.05,
                 verbose = TRUE,
                 OrgDb = org.Hs.eg.db,
                 pAdjustMethod = "BH")

write.csv(as.data.frame(GSEA_result), "../res/DKO_WT_DEGs_GSEA.csv")
```


## 3. Extract key genes from previous work

### Extract genes of interest:

Signature from email on 2025-03-20:

-General cholangiocyte (CO): KRT-7, KRT-19, GGT, SOX9, HNF1B, CFTR 
-Intrahepatic ducts (IHD): DCDC2, PKHD1, HES4, BICC1, JAG1, SOX4, TACSTD2, HES1 
-Common Bile duct (CBD): TFF3, TFF1, TFF2, AQP5
-Gallbladder (GB): FGF19, SLC15A1, MUC13, SOX17, SLC13A1

```{r create gene list 1}
CO_list_1 = c("KRT7", "KRT19", "GGT1", "SOX9", "HNF1B", "CFTR") # GGT1 instead GGT 
IHD_list_1 = c("DCDC2", "PKHD1", "HES4", "BICC1", "JAG1", "SOX4", "TACSTD2", "HES1") 
CBD_list_1 = c("TFF2", "TFF3", "TFF1","AQP5")
GB_list_1 = c("FGF19", "SLC15A1", "MUC13", "SLC13A1", "SOX17")
```

Subset matrix 
```{r Create matrix with key signature}
# scaled_logTMM = scale(logTMM)
reduced_logTMM = logTMM[rownames(logTMM) %in% c(CO_list_1,
                                                IHD_list_1,
                                                CBD_list_1,
                                                GB_list_1),]
reduced_logTMM = reshape::melt(data.table::data.table(reduced_logTMM, 
                                                keep.rownames = TRUE) , 
                         id.vars = c("rn"))
colnames(reduced_logTMM) = c("gene","Sample_ID","value")

# Assigning gene group
reduced_logTMM$Group = ifelse(reduced_logTMM$gene %in% CO_list_1, "General\nCO",
                                         ifelse(reduced_logTMM$gene %in% IHD_list_1, "IHD",
                                                ifelse(reduced_logTMM$gene %in% CBD_list_1, "CBD","GB")))
reduced_logTMM$Group = factor(reduced_logTMM$Group,
                               levels = c("General\nCO", "IHD", "CBD", "GB"))

# Assign condition
reduced_logTMM$Target = factor( unname(condition_dict[reduced_logTMM$Sample_ID]),
                                levels = c("WT","DKO"))

```

```{r Plot heatmap with general CO only, fig.width = 3, fig.height = 4.5}
reduced_logTMM_n = reduced_logTMM
reduced_logTMM_n$Sample_ID = as.character(unname(setNames(sample_data$Sample,sample_data$File_name)[reduced_logTMM_n$Sample_ID]))
reduced_logTMM_n$Sample_ID = unname(setNames(c("Line 1","Line 2","Line 3"),
                            c("325","299","808"))[reduced_logTMM_n$Sample_ID])
reduced_logTMM_n$Target = ifelse(reduced_logTMM_n$Target == "WT", "PCO", "ePCO")
xo = rep(c("atop(atop(textstyle('ePCO'),textstyle(italic('in vitro'))))",
             "atop(atop(textstyle('PCO'),textstyle(italic('in vitro'))))"),3)

reduced_logTMM_n = reduced_logTMM_n[reduced_logTMM_n$Group == "General\nCO",]
  
p <- ggplot(reduced_logTMM_n, aes(gene, Target, fill = value)) +
         geom_tile(colour = "black")

p + viridis::scale_fill_viridis(direction = 1,
                                limits = c(0, 5.6),
  option = "magma",
  na.value = 'black') +
  facet_grid(Sample_ID~., scales = "free", space = "free",switch = "y")+ 
  labs(y=NULL, x=NULL, fill="log(TMM)") +
  theme_bw() +
  # scale_x_discrete(labels=parse(text=xo)) +
  xlab("") + ylab("") +
  theme(panel.background = element_rect(fill = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        panel.grid = element_line(color = "black"),
        strip.background = element_rect(fill="white"),
        legend.position = "top",
        # axis.text.y = element_text(vjust = 0.7),
        legend.box.spacing = unit(-0.5, "pt")
        ) +
  scale_x_discrete(
    expand = expansion(mult = c(0,0)),
    position = "top"
  ) + scale_y_discrete(position = "right")
```


```{r Export figure data}
write.csv(reduced_logTMM_n, "InVitro_Expression.csv")
```


```{r Function to generate split violin}
GeomSplitViolin <- ggproto("GeomSplitViolin", GeomViolin, 
                           draw_group = function(self, data, ..., draw_quantiles = NULL) {
  data <- transform(data, xminv = x - violinwidth * (x - xmin), xmaxv = x + violinwidth * (xmax - x))
  grp <- data[1, "group"]
  newdata <- plyr::arrange(transform(data, x = if (grp %% 2 == 1) xminv else xmaxv), if (grp %% 2 == 1) y else -y)
  newdata <- rbind(newdata[1, ], newdata, newdata[nrow(newdata), ], newdata[1, ])
  newdata[c(1, nrow(newdata) - 1, nrow(newdata)), "x"] <- round(newdata[1, "x"])

  if (length(draw_quantiles) > 0 & !scales::zero_range(range(data$y))) {
    stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <=
      1))
    quantiles <- ggplot2:::create_quantile_segment_frame(data, draw_quantiles)
    aesthetics <- data[rep(1, nrow(quantiles)), setdiff(names(data), c("x", "y")), drop = FALSE]
    aesthetics$alpha <- rep(1, nrow(quantiles))
    both <- cbind(quantiles, aesthetics)
    quantile_grob <- GeomPath$draw_panel(both, ...)
    ggplot2:::ggname("geom_split_violin", grid::grobTree(GeomPolygon$draw_panel(newdata, ...), quantile_grob))
  }
  else {
    ggplot2:::ggname("geom_split_violin", GeomPolygon$draw_panel(newdata, ...))
  }
})

geom_split_violin <- function(mapping = NULL, data = NULL, stat = "ydensity", position = "identity", ..., 
                              draw_quantiles = NULL, trim = TRUE, scale = "area", na.rm = FALSE, 
                              show.legend = NA, inherit.aes = TRUE) {
  layer(data = data, mapping = mapping, stat = stat, geom = GeomSplitViolin, 
        position = position, show.legend = show.legend, inherit.aes = inherit.aes, 
        params = list(trim = trim, scale = scale, draw_quantiles = draw_quantiles, na.rm = na.rm, ...))
}
```


```{r Generate grouped violin plot, fig.width = 3, fig.height = 4}
reduced_scaled_logTMM = scale(logTMM)
reduced_scaled_logTMM = reduced_scaled_logTMM[rownames(reduced_scaled_logTMM) %in% c(CO_list_1,
                                                IHD_list_1,
                                                CBD_list_1,
                                                GB_list_1),]
reduced_scaled_logTMM = reshape::melt(data.table::data.table(reduced_scaled_logTMM, 
                                                keep.rownames = TRUE) , 
                         id.vars = c("rn"))
colnames(reduced_scaled_logTMM) = c("gene","Sample_ID","value")

# Assigning gene group
reduced_scaled_logTMM$Group = ifelse(reduced_scaled_logTMM$gene %in% CO_list_1, "General\nCO",
                                         ifelse(reduced_scaled_logTMM$gene %in% IHD_list_1, "IHD",
                                                ifelse(reduced_scaled_logTMM$gene %in% CBD_list_1, "CBD","GB")))
reduced_scaled_logTMM$Group = factor(reduced_scaled_logTMM$Group,
                               levels = c("General\nCO", "IHD", "CBD", "GB"))

# Assign condition
reduced_scaled_logTMM$Target = factor( unname(condition_dict[reduced_scaled_logTMM$Sample_ID]),
                                levels = c("WT","DKO"))

reduced_logTMM_woGCO = reduced_scaled_logTMM[!reduced_scaled_logTMM$Group== "General\nCO",]
reduced_logTMM_woGCO$Target = ifelse(reduced_logTMM_woGCO$Target == "WT", "PCO","ePCO")


ggplot(reduced_logTMM_woGCO, aes(Group, value, fill = Target)) + 
  geom_split_violin() +
  geom_jitter(aes(colour = Target), 
     position = position_jitterdodge(), alpha = 0.6) +
  xlab("") + ylab("z-score") +
  coord_flip() +
  theme_bw() +
  scale_fill_manual(values=c("PCO" = "#B9B9FF","ePCO" =  "#FEB4FF")) + 
  theme(legend.position="bottom") +
  scale_color_manual(values = c("PCO" = "purple",
                                "ePCO" = "hotpink"))
  
```

```{r Export results}
write.csv(reduced_logTMM_woGCO,"InVitro_Zscore_Expression.csv")
```

## 5. Generate  barplot with in vitro data

```{r Read barplot data from previous analysis}
bar_all_data_CO = read.csv("../../DKO_Analysis/res/ExpressionSummary/ExpressionSummary_010525/BarPlot_Data.csv")
bar_all_data_CO$X = NULL
bar_all_data_CO = bar_all_data_CO[c( "Sample_ID","gene","value","Target","Group")]

invitro_CO = reduced_logTMM[reduced_logTMM$Group == "General\nCO",]
general_co_data = rbind(bar_all_data_CO, invitro_CO)
general_co_data$Target = gsub("atop(atop(textstyle('PCO'),textstyle('GB')))","atop(atop(textstyle('Prim'),textstyle('GB')))",general_co_data$Target)
general_co_data$Target[general_co_data$Target == "WT"] = "atop(atop(textstyle('PCO'),textstyle(italic('in vitro'))))" 
general_co_data$Target[general_co_data$Target == "DKO"] = "atop(atop(textstyle('ePCO'),textstyle(italic('in vitro'))))" 
general_co_data$Target = factor(general_co_data$Target,
                                levels = c("atop(atop(textstyle('PCO'),textstyle(italic('in vitro'))))",
                                           "atop(atop(textstyle('ePCO'),textstyle(italic('in vitro'))))",
                                           "atop(atop(textstyle('PCO'),textstyle(italic('in vivo'))))",
                                           "atop(atop(textstyle('ePCO'),textstyle(italic('in vivo'))))",
                                           "atop(atop(textstyle('Prim'),textstyle('IHD')))",
                                           "atop(atop(textstyle('Prim'),textstyle('CBD')))",
                                           "atop(atop(textstyle('Prim'),textstyle('GB')))" ))

ggplot(general_co_data, aes( y = value, x = Target), fill = "#7D7D7D", color = "#7D7D7D") +
  geom_bar(
    position = "dodge", width = 0.8, stat = "summary", fun = "mean",
    color = "#7D7D7D", fill = "#7D7D7D", linewidth = .8
  ) +
  geom_point(colour = "black", fill = "white",
    alpha = 0.3
  ) + 
  stat_summary(fun.ymin = function(x) mean(x) - sd(x), 
           fun.ymax = function(x) mean(x) + sd(x), position ='dodge', 
           geom = 'errorbar', color = "black",
            width = 0.2, linewidth = 0.8,  na.rm = TRUE) + 
  theme_classic() +
  theme(legend.position = "None") +
  labs(y="Mean Expresion (log(TMM))", x=NULL, fill="#7D7D7D") +
  theme(aspect.ratio = 5/4) +
      stat_summary(aes(label=round(..y..,2)), fun.y=mean, geom="text", size=4,
             vjust = -6,  na.rm = TRUE) +
  scale_y_continuous(expand =  expansion(mult = c(0, 0), 
                                         add = c(0, 1)),
                     limits = c(0, 12)) +
  scale_x_discrete(labels = parse(text = levels(general_co_data$Target))) 
```

```{r Export data results}
write.csv(general_co_data, "AllExpression_GeneralCOSummary.csv")
```


Session information 
```{r}
sessionInfo()
```








